ball_finder.py

```diff
@@ -0,0 +1,81 @@
+import cv2
+import math
+import numpy as np
+
+
+class BallFinder:
+    def __init__(self, detector):
+        self.detector = detector
+        self.generator = detector.get_generator()
+
+        frame, _ = self.generator.generate()
+        self.SCREEN_HEIGHT, self.SCREEN_WIDTH = frame.shape[:2]
+        self.FOCAL_LENGTH_PIXELS = 657
+
+        # Depth csv of empty floor
+        self.floor_frame = np.loadtxt("FLOOR.csv", dtype=np.float32, delimiter=",")
+        self.floor_frame[self.floor_frame > 4] = 0
+
+        # Displacement constants between rgb frame and depth frame
+        self.DEPTH_X_SHIFT = -23
+        self.DEPTH_Y_SHIFT = 31
+
+        # Height of the camera (in m)
+        self.CAMERA_HEIGHT = 0.7
+
+        # Minimum distance (in m) for finding obstacles (the distance from the camera to the front of the robot)
+        self.MIN_OBSTACLE_DIST = 0.56
+
+    # Returns depth frame where all floor points have depth 0.
+    def remove_floor(self, depth):
+        return np.where(abs(depth - self.floor_frame) < 0.05, 0, depth)
+
+    # Returns the center (x, y) of a contour based on its bounding box.
+    def get_contour_center(self, contour):
+        x, y, w, h = cv2.boundingRect(contour)
+        return x + w / 2, y + h / 2
+
+    # Returns horizontally straight distance (in m) to a ball or obstacle.
+    # is_ball is used to determine whether or not to shift the x and y positions the amount needed when using the depth
+    # camera.
+    def get_distance(self, depth_frame, item, is_ball=False):
+        x = min(639, int(item[0] + is_ball * self.DEPTH_X_SHIFT))
```

Comment:
Why would only balls need shifting?

Reply:
Inside `PowerCellDetector`, the ball x and y are found using the color frame so we need to shift when accessing depth. We find obstacles directly from the depth frame.